kamu-core/src/infra/repos/metadata_chain_impl.rs

```diff
@@ -156,34 +188,48 @@ where
     }
 
     async fn get_block(&self, hash: &Multihash) -> Result<MetadataBlock, GetBlockError> {
-        let data = match self.obj_repo.get_bytes(hash).await {
+        let data = self.get_block_bytes(hash).await?;
+
+        match self.construct_block_from_bytes(hash, data).await {
+            Ok(block) => Ok(block),
+            Err(e) => match e {
+                ConstructBlockFromBytesError::BlockVersion(e) => Err(GetBlockError::BlockVersion(e)),
+                ConstructBlockFromBytesError::Internal(e) => Err(GetBlockError::Internal(e)),
+            }
+        }
+    }
+
+    async fn get_block_size(&self, hash: &Multihash) -> Result<u64, GetBlockError> {
+        let size_result = self.obj_repo.get_size(hash).await;
+        match size_result {
+            Ok(size) => Ok(size),
+            Err(GetError::NotFound(e)) => {
+                Err(GetBlockError::NotFound(BlockNotFoundError { hash: e.hash }))
+            }
+            Err(GetError::Access(e)) => Err(GetBlockError::Access(e)),
+            Err(GetError::Internal(e)) => Err(GetBlockError::Internal(e)),
+        }
+    }
+
+    async fn get_block_bytes(&self, hash: &Multihash) -> Result<Bytes, GetBlockError> {
+        match self.obj_repo.get_bytes(hash).await {
             Ok(data) => Ok(data),
             Err(GetError::NotFound(e)) => {
                 Err(GetBlockError::NotFound(BlockNotFoundError { hash: e.hash }))
             }
             Err(GetError::Access(e)) => Err(GetBlockError::Access(e)),
             Err(GetError::Internal(e)) => Err(GetBlockError::Internal(e)),
-        }?;
+        }
+    }
 
-        match FlatbuffersMetadataBlockDeserializer.read_manifest(&data) {
-            Ok(block) => return Ok(block),
-            Err(e) => match e {
-                Error::UnsupportedVersion { .. } => {
-                    return Err(GetBlockError::BlockVersion(BlockVersionError {
-                        hash: hash.clone(),
-                        source: e.into(),
-                    }));
-                }
-                _ => {
-                    return Err(GetBlockError::Internal(
-                        BlockMalformedError {
-                            hash: hash.clone(),
-                            source: e.into(),
-                        }
-                        .int_err(),
-                    ));
-                }
-            },
+    async fn get_block_download_url(&self, prefix_url: &Url, hash: &Multihash) -> Result<(Url, Option<DateTime<Utc>>), GetBlockError> {
+        match self.obj_repo.get_download_url(prefix_url, hash).await {
```

Comment:
Can you explain `prefix_url` here? I'm confused why it's needed as object repos should know exactly where data is.

Reply:
With S3 repository, the output URL depends solely on the external storage system + bucket key.
With local repository, the repository can only tell the relative URL part, but it has no idea on the API server context (host+port), nor the dataset context (author?+name).

Reply:
Suggested a solution below - will resolve this comment.